File: src/backend-api.Api/Services/TransactionService.cs
using backend_api.Api.DTOs;
using backend_api.Api.Repositories;

namespace backend_api.Api.Services;

public class TransactionService : ITransactionService
{
    private readonly ITransactionRepository _transactionRepo;

    public TransactionService(ITransactionRepository transactionRepo)
    {
        _transactionRepo = transactionRepo;
    }

    public async Task<IEnumerable<TransactionResponse>> GetMyTransactionsAsync(string userId, string? transType = null)
    {
        var transactions = await _transactionRepo.GetByUserIdAsync(userId, transType);

        return transactions.Select(t => new TransactionResponse
        {
            TransId = t.TransId,
            RefId = t.RefId,
            TransType = t.TransType,
            Amount = t.Amount,
            BalanceBefore = t.BalanceBefore,
            BalanceAfter = t.BalanceAfter,
            Description = t.Description,
            TransTime = t.TransTime
        });
    }
}

File: src/frontend-web/src/app/core/services/transaction.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface TransactionResponse {
    transId: number;
    refId: string;
    transType: string;
    amount: number;
    balanceBefore: number;
    balanceAfter: number;
    description: string;
    transTime: string;
}

@Injectable({ providedIn: 'root' })
export class TransactionService {
    private readonly http = inject(HttpClient);
    private readonly apiUrl = `${environment.apiUrl}/transactions`;

    getMyTransactions(type?: string): Observable<TransactionResponse[]> {
        const url = type ? `${this.apiUrl}?type=${type}` : this.apiUrl;
        return this.http.get<TransactionResponse[]>(url);
    }
}
